src/sagas/tokens.js

```diff
@@ -0,0 +1,255 @@
+  } catch (e) {
+    yield put(tokenFetchHistoryFailed(tokenId));
+  }
+}
+
+/**
+ * This saga will route the actions dispatched from SET_TOKEN and NEW_TOKEN to the
+ * TOKEN_FETCH_BALANCE_REQUESTED saga, the idea is to load the balance for new tokens
+ * registered or created on the app.
+ */
+function* routeTokenChange(action) {
+  const wallet = yield select((state) => state.wallet);
+
+  if (!wallet || !wallet.isReady()) {
+    return;
+  }
+
+  switch (action.type) {
+    default:
+    case 'SET_TOKENS':
+      for (const token of action.payload) {
+        yield put({ type: types.TOKEN_FETCH_BALANCE_REQUESTED, tokenId: token.uid });
+      }
+      break;
+    case 'NEW_TOKEN':
+      yield put({ type: types.TOKEN_FETCH_HISTORY_REQUESTED, tokenId: action.payload.uid });
+      break;
+  }
+}
+
+/**
+ * This saga will create a channel to queue TOKEN_FETCH_METADATA_REQUESTED actions and
+ * consumers that will run in parallel consuming those actions.
+ *
+ * More information about channels can be read on https://redux-saga.js.org/docs/api/#takechannel
+ */
+function* fetchTokenMetadataQueue() {
+  const fetchTokenMetadataChannel = yield call(channel);
+
+  // Fork CONCURRENT_FETCH_REQUESTS threads to download token balances
+  for (let i = 0; i < CONCURRENT_FETCH_REQUESTS; i += 1) {
+    yield fork(fetchTokenMetadataConsumer, fetchTokenMetadataChannel);
+  }
+
+  while (true) {
+    const action = yield take(types.TOKEN_FETCH_METADATA_REQUESTED);
+    yield put(fetchTokenMetadataChannel, action);
+  }
+}
+
+/**
+ * This saga will consume the fetchTokenBalanceChannel for TOKEN_FETCH_BALANCE_REQUEST actions
+ * and wait until the TOKEN_FETCH_BALANCE_SUCCESS action is dispatched with the specific tokenId
+ */
+function* fetchTokenMetadataConsumer(fetchTokenMetadataChannel) {
+  while (true) {
+    const action = yield take(fetchTokenMetadataChannel);
+
+    yield fork(fetchTokenMetadata, action);
+
+    // Wait until the success action is dispatched before consuming another action
+    yield take(
+      specificTypeAndPayload([
+        types.TOKEN_FETCH_METADATA_SUCCESS,
+        types.TOKEN_FETCH_METADATA_FAILED,
+      ], {
+        tokenId: action.tokenId,
+      }),
+    );
+  }
+}
+
+/**
+ * Fetch a single token from the metadataApi
+ *
+ * @param {Array} token The token to fetch from the metadata api
+ * @param {String} network Network name
+ *
+ * @memberof Wallet
+ * @inner
+ */
+export function* fetchTokenMetadata({ tokenId }) {
+  const { network } = yield select((state) => state.serverInfo);
+
+  try {
+    // Retry mechanism
+    for (let i = 0; i <= METADATA_MAX_RETRIES; i += 1) {
+      try {
+        const data = yield call(metadataApi.getDagMetadata, tokenId, network);
+
+        yield put({
+          type: types.TOKEN_FETCH_METADATA_SUCCESS,
+          tokenId,
+          data: get(data, tokenId, null),
+        });
+        return;
+      } catch (e) {
+        yield delay(1000); // Wait 1s before trying again
```

Comment:
I guess that failures here are expected since this is a http request, we don't need to get notified about it

If all retries fail, we already have a log indicating on line 239